File: tools/marvin/marvin/integration/lib/base/DiskOffering.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
from marvin.integration.lib.base import CloudStackEntity
from marvin.cloudstackAPI import createDiskOffering
from marvin.cloudstackAPI import listDiskOfferings
from marvin.cloudstackAPI import updateDiskOffering
from marvin.cloudstackAPI import deleteDiskOffering

class DiskOffering(CloudStackEntity.CloudStackEntity):


    def __init__(self, items):
        self.__dict__.update(items)


    @classmethod
    def create(cls, apiclient, factory, **kwargs):
        cmd = createDiskOffering.createDiskOfferingCmd()
        [setattr(cmd, factoryKey, factoryValue) for factoryKey, factoryValue in factory.__dict__.iteritems()]
        [setattr(cmd, key, value) for key,value in kwargs.iteritems()]
        diskoffering = apiclient.createDiskOffering(cmd)
        return DiskOffering(diskoffering.__dict__)


    @classmethod
    def list(self, apiclient, **kwargs):
        cmd = listDiskOfferings.listDiskOfferingsCmd()
        [setattr(cmd, key, value) for key,value in kwargs.iteritems()]
        diskoffering = apiclient.listDiskOfferings(cmd)
        return map(lambda e: DiskOffering(e.__dict__), diskoffering)


    def update(self, apiclient, **kwargs):
        cmd = updateDiskOffering.updateDiskOfferingCmd()
        cmd.id = self.id
        [setattr(cmd, key, value) for key,value in kwargs.iteritems()]
        diskoffering = apiclient.updateDiskOffering(cmd)
        return diskoffering


    def delete(self, apiclient, **kwargs):
        cmd = deleteDiskOffering.deleteDiskOfferingCmd()
        cmd.id = self.id
        [setattr(cmd, key, value) for key,value in kwargs.iteritems()]
        diskoffering = apiclient.deleteDiskOffering(cmd)
        return diskoffering

File: api/src/main/java/com/cloud/offering/DiskOffering.java
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.
package com.cloud.offering;

import java.util.Date;

import org.apache.cloudstack.acl.InfrastructureEntity;
import org.apache.cloudstack.api.Identity;
import org.apache.cloudstack.api.InternalIdentity;

import com.cloud.storage.Storage.ProvisioningType;

/**
 * Represents a disk offering that specifies what the end user needs in
 * the disk offering.
 *
 */
public interface DiskOffering extends InfrastructureEntity, Identity, InternalIdentity {
    enum State {
        Inactive, Active,
    }

    State getState();

    enum DiskCacheMode {
        NONE("none"), WRITEBACK("writeback"), WRITETHROUGH("writethrough");

        private final String _diskCacheMode;

        DiskCacheMode(String cacheMode) {
            _diskCacheMode = cacheMode;
        }

        @Override
        public String toString() {
            return _diskCacheMode;
        }
    };

    String getUniqueName();

    boolean isUseLocalStorage();

    String getName();

    boolean isSystemUse();

    String getDisplayText();

    ProvisioningType getProvisioningType();

    String getTags();

    String[] getTagsArray();

    Date getCreated();

    boolean isCustomized();

    void setDiskSize(long diskSize);

    long getDiskSize();

    void setCustomizedIops(Boolean customizedIops);

    Boolean isCustomizedIops();

    void setMinIops(Long minIops);

    Long getMinIops();

    void setMaxIops(Long maxIops);

    Long getMaxIops();

    boolean isRecreatable();

    void setBytesReadRate(Long bytesReadRate);

    Long getBytesReadRate();

    void setBytesReadRateMax(Long bytesReadRateMax);

    Long getBytesReadRateMax();

    void setBytesReadRateMaxLength(Long bytesReadRateMaxLength);

    Long getBytesReadRateMaxLength();


    void setBytesWriteRate(Long bytesWriteRate);

    Long getBytesWriteRate();

    void setBytesWriteRateMax(Long bytesWriteMax);

    Long getBytesWriteRateMax();

    void setBytesWriteRateMaxLength(Long bytesWriteMaxLength);

    Long getBytesWriteRateMaxLength();


    void setIopsReadRate(Long iopsReadRate);

    Long getIopsReadRate();

    void setIopsReadRateMax(Long iopsReadRateMax);

    Long getIopsReadRateMax();

    void setIopsReadRateMaxLength(Long iopsReadRateMaxLength);

    Long getIopsReadRateMaxLength();

    void setIopsWriteRate(Long iopsWriteRate);

    Long getIopsWriteRate();

    void setIopsWriteRateMax(Long iopsWriteRateMax);

    Long getIopsWriteRateMax();


    void setIopsWriteRateMaxLength(Long iopsWriteRateMaxLength);

    Long getIopsWriteRateMaxLength();

    void setHypervisorSnapshotReserve(Integer hypervisorSnapshotReserve);

    Integer getHypervisorSnapshotReserve();

    DiskCacheMode getCacheMode();

    void setCacheMode(DiskCacheMode cacheMode);

    boolean isComputeOnly();
}
